File: netlify/functions/webflow_proxy.js
import fetch from 'node-fetch';

const API_TOKEN = process.env.WEBFLOW_API_TOKEN;
const API_BASE_URL = "https://api.webflow.com/v2";
const COLLECTION_IDS = {
    "media_assets": "6891479d29ed1066b71124e9",
    "voxpro_assignments": "689ac6bdf10259dd9be04e16"
};

const CORS_HEADERS = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'POST, OPTIONS'
};

export const handler = async (event ) => {
    if (event.httpMethod === 'OPTIONS' ) {
        return { statusCode: 204, headers: CORS_HEADERS, body: '' };
    }
    if (event.httpMethod !== 'POST' ) {
        return { statusCode: 405, headers: CORS_HEADERS, body: JSON.stringify({ error: 'Method Not Allowed' }) };
    }

    try {
        const { endpoint, method, body } = JSON.parse(event.body);
        const [collection_key, item_id] = endpoint.split('/');
        const collection_id = COLLECTION_IDS[collection_key];

        if (!collection_id) throw new Error(`Invalid endpoint: ${collection_key}`);

        const url = item_id 
            ? `${API_BASE_URL}/collections/${collection_id}/items/${item_id}`
            : `${API_BASE_URL}/collections/${collection_id}/items`;

        const options = {
            method: method || 'GET',
            headers: {
                'Authorization': `Bearer ${API_TOKEN}`,
                'accept': 'application/json',
                'Content-Type': 'application/json'
            }
        };

        // **THE DEFINITIVE, FINAL FIX**
        // The entire payload, including all custom fields, MUST be inside a `fieldData` object.
        if (body && (method === 'POST' || method === 'PATCH')) {
            options.body = JSON.stringify({ fieldData: body.fieldData });
        }

        const response = await fetch(url, options);
        const responseText = await response.text();
        
        if (!response.ok) {
            throw new Error(`Webflow API Error (${response.status}): ${responseText}`);
        }
        
        const finalBody = responseText || JSON.stringify({ success: true });
        return { statusCode: 200, headers: CORS_HEADERS, body: finalBody };

    } catch (error) {
        console.error("Handler Error:", error);
        return { statusCode: 500, headers: CORS_HEADERS, body: JSON.stringify({ error: error.message }) };
    }
};

File: routes/webflow_proxy.py
import requests
from flask import Blueprint, request, jsonify
import os
import json

webflow_proxy_blueprint = Blueprint('webflow_proxy', __name__)

# --- FINAL CONFIGURATION ---
API_TOKEN = "aa08be8443f9e389f9c21e1fb6f4566517403ad131b31a7f3c1467ba49155c35"
SITE_ID = "688ed8debc05764047afa2a7"
COLLECTION_IDS = {
    "media_assets": "6891479d29ed1066b71124e9",
    "voxpro_assignments": "689ac6bdf10259dd9be04e16"
}
# --- END CONFIGURATION ---

API_BASE_URL = "https://api.webflow.com/v2"
HEADERS = {
    "accept": "application/json",
    "authorization": f"Bearer {API_TOKEN}"
}

@webflow_proxy_blueprint.route('/webflow_proxy', methods=['POST', 'GET'] )
def proxy():
    # Health check endpoint for Render
    if request.method == 'GET':
        return jsonify({"status": "ok", "message": "Proxy is running"}), 200

    # Main proxy logic for POST requests
    try:
        data = request.get_json()
        if not data:
            return jsonify({"error": "Invalid JSON payload"}), 400

        endpoint_key = data.get('endpoint')
        method = data.get('method', 'GET').upper()
        body = data.get('body')
        url = ""

        print(f"Received request: Method={method}, Endpoint={endpoint_key}")

        if not endpoint_key:
            return jsonify({"error": "Endpoint key is missing"}), 400

        if '/' in endpoint_key:
            collection_key, item_id = endpoint_key.split('/', 1)
            collection_id = COLLECTION_IDS.get(collection_key)
            if not collection_id:
                return jsonify({"error": f"Invalid endpoint key: {collection_key}"}), 400
            url = f"{API_BASE_URL}/collections/{collection_id}/items/{item_id}"
        else:
            collection_id = COLLECTION_IDS.get(endpoint_key)
            if not collection_id:
                return jsonify({"error": f"Invalid endpoint key: {endpoint_key}"}), 400
            url = f"{API_BASE_URL}/collections/{collection_id}/items"

        req_headers = HEADERS.copy()
        if method in ['POST', 'PATCH']:
            req_headers['Content-Type'] = 'application/json'

        print(f"Proxying request to URL: {url}")
        response = requests.request(method, url, headers=req_headers, json=body, timeout=20)
        response.raise_for_status()

        # Handle successful but empty responses (e.g., DELETE or empty collections)
        if response.status_code == 204 or not response.content:
            print("Request successful with no content.")
            # For GET requests on an empty collection, Webflow returns a specific structure
            if method == 'GET' and not '/' in endpoint_key:
                 return jsonify({"items": [], "count": 0, "limit": 100, "offset": 0, "total": 0}), 200
            return jsonify({"success": True, "message": "Operation successful"}), 200

        print("Request successful with content.")
        return jsonify(response.json()), response.status_code

    except requests.exceptions.Timeout:
        print("Error: Request to Webflow API timed out.")
        return jsonify({"error": "The request to Webflow API timed out."}), 504
    except requests.exceptions.RequestException as e:
        error_message = f"API request failed: {e}"
        if e.response is not None:
            error_message += f" - Details: {e.response.text}"
        print(f"Error: {error_message}")
        return jsonify({"error": error_message}), 500
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        return jsonify({"error": "An unexpected server error occurred."}), 500

